File: app/src/main/java/acr/browser/lightning/browser/fragment/BookmarksFragment.java
package acr.browser.lightning.browser.fragment;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.PorterDuff;
import android.os.Bundle;
import android.support.annotation.IdRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewCompat;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import com.anthonycr.bonsai.Schedulers;
import com.anthonycr.bonsai.SingleOnSubscribe;
import com.anthonycr.bonsai.Subscription;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import javax.inject.Inject;

import acr.browser.lightning.R;
import acr.browser.lightning.browser.bookmark.BookmarkUiModel;
import acr.browser.lightning.reading.activity.ReadingActivity;
import acr.browser.lightning.browser.TabsManager;
import acr.browser.lightning.animation.AnimationUtils;
import acr.browser.lightning.BrowserApp;
import acr.browser.lightning.browser.BookmarksView;
import acr.browser.lightning.constant.Constants;
import acr.browser.lightning.controller.UIController;
import acr.browser.lightning.database.HistoryItem;
import acr.browser.lightning.database.bookmark.BookmarkModel;
import acr.browser.lightning.dialog.LightningDialogBuilder;
import acr.browser.lightning.favicon.FaviconModel;
import acr.browser.lightning.preference.PreferenceManager;
import acr.browser.lightning.utils.Preconditions;
import acr.browser.lightning.utils.SubscriptionUtils;
import acr.browser.lightning.utils.ThemeUtils;
import acr.browser.lightning.view.LightningView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

public class BookmarksFragment extends Fragment implements View.OnClickListener, View.OnLongClickListener, BookmarksView {

    @NonNull
    public static BookmarksFragment createFragment(boolean isIncognito) {
        BookmarksFragment bookmarksFragment = new BookmarksFragment();
        final Bundle bookmarksFragmentArguments = new Bundle();
        bookmarksFragmentArguments.putBoolean(BookmarksFragment.INCOGNITO_MODE, isIncognito);
        bookmarksFragment.setArguments(bookmarksFragmentArguments);

        return bookmarksFragment;
    }

    private static final String TAG = "BookmarksFragment";

    private final static String INCOGNITO_MODE = TAG + ".INCOGNITO_MODE";

    // Managers
    @Inject BookmarkModel mBookmarkManager;

    // Dialog builder
    @Inject LightningDialogBuilder mBookmarksDialogBuilder;

    @Inject PreferenceManager mPreferenceManager;

    @Inject FaviconModel mFaviconModel;

    private TabsManager mTabsManager;

    private UIController mUiController;

    // Adapter
    private BookmarkListAdapter mBookmarkAdapter;

    // Preloaded images
    private Bitmap mWebpageBitmap, mFolderBitmap;

    // Views
    @BindView(R.id.right_drawer_list) RecyclerView mBookmarksListView;
    @BindView(R.id.starIcon) ImageView mBookmarkTitleImage;
    @BindView(R.id.icon_star) ImageView mBookmarkImage;

    @Nullable private Unbinder mUnbinder;

    // Colors
    private int mIconColor, mScrollIndex;

    private boolean mIsIncognito;

    @Nullable private Subscription mBookmarksSubscription;
    @Nullable private Subscription mFoldersSubscription;
    @Nullable private Subscription mBookmarkUpdateSubscription;

    @NonNull private final BookmarkUiModel mUiModel = new BookmarkUiModel();

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        BrowserApp.getAppComponent().inject(this);
        final Bundle arguments = getArguments();
        final Context context = getContext();
        mUiController = (UIController) context;
        mTabsManager = mUiController.getTabModel();
        mIsIncognito = arguments.getBoolean(INCOGNITO_MODE, false);
        boolean darkTheme = mPreferenceManager.getUseTheme() != 0 || mIsIncognito;
        mWebpageBitmap = ThemeUtils.getThemedBitmap(context, R.drawable.ic_webpage, darkTheme);
        mFolderBitmap = ThemeUtils.getThemedBitmap(context, R.drawable.ic_folder, darkTheme);
        mIconColor = darkTheme ? ThemeUtils.getIconDarkThemeColor(context) :
            ThemeUtils.getIconLightThemeColor(context);
    }

    private TabsManager getTabsManager() {
        if (mTabsManager == null) {
            mTabsManager = mUiController.getTabModel();
        }
        return mTabsManager;
    }

    // Handle bookmark click
    private final OnItemClickListener mItemClickListener = new OnItemClickListener() {
        @Override
        public void onItemClick(@NonNull HistoryItem item) {
            if (item.isFolder()) {
                mScrollIndex = ((LinearLayoutManager) mBookmarksListView.getLayoutManager()).findFirstVisibleItemPosition();
                setBookmarksShown(item.getTitle(), true);
            } else {
                mUiController.bookmarkItemClicked(item);
            }
        }
    };

    private final OnItemLongClickListener mItemLongClickListener = new OnItemLongClickListener() {
        @Override
        public boolean onItemLongClick(@NonNull HistoryItem item) {
            handleLongPress(item);
            return true;
        }
    };

    @Override
    public void onResume() {
        super.onResume();
        if (mBookmarkAdapter != null) {
            setBookmarksShown(null, false);
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.bookmark_drawer, container, false);
        mUnbinder = ButterKnife.bind(this, view);
        mBookmarkTitleImage.setColorFilter(mIconColor, PorterDuff.Mode.SRC_IN);
        final View backView = view.findViewById(R.id.bookmark_back_button);
        backView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (!mUiModel.isRootFolder()) {
                    setBookmarksShown(null, true);
                    mBookmarksListView.getLayoutManager().scrollToPosition(mScrollIndex);
                }
            }
        });
        setupNavigationButton(view, R.id.action_add_bookmark, R.id.icon_star);
        setupNavigationButton(view, R.id.action_reading, R.id.icon_reading);
        setupNavigationButton(view, R.id.action_toggle_desktop, R.id.icon_desktop);

        mBookmarkAdapter = new BookmarkListAdapter(mFaviconModel, mFolderBitmap, mWebpageBitmap);
        mBookmarkAdapter.setOnItemClickListener(mItemClickListener);
        mBookmarkAdapter.setOnItemLongClickListener(mItemLongClickListener);
        mBookmarksListView.setLayoutManager(new LinearLayoutManager(getContext()));
        mBookmarksListView.setAdapter(mBookmarkAdapter);

        setBookmarksShown(null, true);

        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

        SubscriptionUtils.safeUnsubscribe(mBookmarksSubscription);
        SubscriptionUtils.safeUnsubscribe(mFoldersSubscription);
        SubscriptionUtils.safeUnsubscribe(mBookmarkUpdateSubscription);

        if (mBookmarkAdapter != null) {
            mBookmarkAdapter.cleanupSubscriptions();
        }

        if (mUnbinder != null) {
            mUnbinder.unbind();
            mUnbinder = null;
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        SubscriptionUtils.safeUnsubscribe(mBookmarksSubscription);
        SubscriptionUtils.safeUnsubscribe(mFoldersSubscription);
        SubscriptionUtils.safeUnsubscribe(mBookmarkUpdateSubscription);

        if (mBookmarkAdapter != null) {
            mBookmarkAdapter.cleanupSubscriptions();
        }
    }

    public void reinitializePreferences() {
        Activity activity = getActivity();
        if (activity == null) {
            return;
        }
        boolean darkTheme = mPreferenceManager.getUseTheme() != 0 || mIsIncognito;
        mWebpageBitmap = ThemeUtils.getThemedBitmap(activity, R.drawable.ic_webpage, darkTheme);
        mFolderBitmap = ThemeUtils.getThemedBitmap(activity, R.drawable.ic_folder, darkTheme);
        mIconColor = darkTheme ? ThemeUtils.getIconDarkThemeColor(activity) :
            ThemeUtils.getIconLightThemeColor(activity);
    }

    private void updateBookmarkIndicator(final String url) {
        SubscriptionUtils.safeUnsubscribe(mBookmarkUpdateSubscription);
        mBookmarkUpdateSubscription = mBookmarkManager.isBookmark(url)
            .subscribeOn(Schedulers.io())
            .observeOn(Schedulers.main())
            .subscribe(new SingleOnSubscribe<Boolean>() {
                @Override
                public void onItem(@Nullable Boolean item) {
                    mBookmarkUpdateSubscription = null;
                    Preconditions.checkNonNull(item);
                    Activity activity = getActivity();
                    if (mBookmarkImage == null || activity == null) {
                        return;
                    }
                    if (!item) {
                        mBookmarkImage.setImageResource(R.drawable.ic_action_star);
                        mBookmarkImage.setColorFilter(mIconColor, PorterDuff.Mode.SRC_IN);
                    } else {
                        mBookmarkImage.setImageResource(R.drawable.ic_bookmark);
                        mBookmarkImage.setColorFilter(ThemeUtils.getAccentColor(activity), PorterDuff.Mode.SRC_IN);
                    }
                }
            });
    }

    @Override
    public void handleBookmarkDeleted(@NonNull HistoryItem item) {
        if (item.isFolder()) {
            setBookmarksShown(null, false);
        } else {
            mBookmarkAdapter.deleteItem(item);
        }
    }

    private void setBookmarksShown(@Nullable final String folder, final boolean animate) {
        SubscriptionUtils.safeUnsubscribe(mBookmarksSubscription);
        mBookmarksSubscription = mBookmarkManager.getBookmarksFromFolderSorted(folder)
            .subscribeOn(Schedulers.io())
            .observeOn(Schedulers.main())
            .subscribe(new SingleOnSubscribe<List<HistoryItem>>() {
                @Override
                public void onItem(@Nullable final List<HistoryItem> item) {
                    mBookmarksSubscription = null;
                    Preconditions.checkNonNull(item);

                    mUiModel.setCurrentFolder(folder);
                    if (folder == null) {
                        SubscriptionUtils.safeUnsubscribe(mFoldersSubscription);
                        mFoldersSubscription = mBookmarkManager.getFoldersSorted()
                            .subscribeOn(Schedulers.io())
                            .observeOn(Schedulers.main())
                            .subscribe(new SingleOnSubscribe<List<HistoryItem>>() {
                                @Override
                                public void onItem(@Nullable List<HistoryItem> folders) {
                                    mFoldersSubscription = null;
                                    Preconditions.checkNonNull(folders);
                                    item.addAll(folders);
                                    setBookmarkDataSet(item, animate);
                                }
                            });
                    } else {
                        setBookmarkDataSet(item, animate);
                    }
                }
            });
    }

    private void setBookmarkDataSet(@NonNull List<HistoryItem> items, boolean animate) {
        mBookmarkAdapter.updateItems(items);
        final int resource;
        if (mUiModel.isRootFolder()) {
            resource = R.drawable.ic_action_star;
        } else {
            resource = R.drawable.ic_action_back;
        }

        if (animate) {
            Animation transition = AnimationUtils.createRotationTransitionAnimation(mBookmarkTitleImage, resource);
            mBookmarkTitleImage.startAnimation(transition);
        } else {
            mBookmarkTitleImage.setImageResource(resource);
        }
    }

    private void setupNavigationButton(@NonNull View view, @IdRes int buttonId, @IdRes int imageId) {
        FrameLayout frameButton = view.findViewById(buttonId);
        frameButton.setOnClickListener(this);
        frameButton.setOnLongClickListener(this);
        ImageView buttonImage = view.findViewById(imageId);
        buttonImage.setColorFilter(mIconColor, PorterDuff.Mode.SRC_IN);
    }

    private void handleLongPress(@NonNull final HistoryItem item) {
        if (item.isFolder()) {
            mBookmarksDialogBuilder.showBookmarkFolderLongPressedDialog(getActivity(), mUiController, item);
        } else {
            mBookmarksDialogBuilder.showLongPressedDialogForBookmarkUrl(getActivity(), mUiController, item);
        }
    }

    @Override
    public void onClick(@NonNull View v) {
        switch (v.getId()) {
            case R.id.action_add_bookmark:
                mUiController.bookmarkButtonClicked();
                break;
            case R.id.action_reading:
                LightningView currentTab = getTabsManager().getCurrentTab();
                if (currentTab != null) {
                    Intent read = new Intent(getActivity(), ReadingActivity.class);
                    read.putExtra(Constants.LOAD_READING_URL, currentTab.getUrl());
                    startActivity(read);
                }
                break;
            case R.id.action_toggle_desktop:
                LightningView current = getTabsManager().getCurrentTab();
                if (current != null) {
                    current.toggleDesktopUA(getActivity());
                    current.reload();
                    // TODO add back drawer closing
                }
                break;
            default:
                break;
        }
    }

    @Override
    public boolean onLongClick(View v) {
        return false;
    }

    @Override
    public void navigateBack() {
        if (mUiModel.isRootFolder()) {
            mUiController.closeBookmarksDrawer();
        } else {
            setBookmarksShown(null, true);
            mBookmarksListView.getLayoutManager().scrollToPosition(mScrollIndex);
        }
    }

    @Override
    public void handleUpdatedUrl(@NonNull String url) {
        updateBookmarkIndicator(url);
        String folder = mUiModel.getCurrentFolder();
        setBookmarksShown(folder, false);
    }

    static class BookmarkViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener, View.OnLongClickListener {
        @BindView(R.id.textBookmark) TextView txtTitle;
        @BindView(R.id.faviconBookmark) ImageView favicon;

        @NonNull private final BookmarkListAdapter adapter;

        @Nullable private final OnItemLongClickListener onItemLongClickListener;
        @Nullable private final OnItemClickListener onItemClickListener;

        BookmarkViewHolder(@NonNull View itemView,
                           @NonNull BookmarkListAdapter adapter,
                           @Nullable OnItemLongClickListener onItemLongClickListener,
                           @Nullable OnItemClickListener onItemClickListener) {
            super(itemView);
            ButterKnife.bind(this, itemView);

            this.adapter = adapter;

            this.onItemClickListener = onItemClickListener;
            this.onItemLongClickListener = onItemLongClickListener;

            itemView.setOnLongClickListener(this);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            int index = getAdapterPosition();
            if (onItemClickListener != null && index != RecyclerView.NO_ID) {
                onItemClickListener.onItemClick(adapter.itemAt(index));
            }
        }

        @Override
        public boolean onLongClick(View v) {
            int index = getAdapterPosition();
            return index != RecyclerView.NO_POSITION && onItemLongClickListener != null &&
                onItemLongClickListener.onItemLongClick(adapter.itemAt(index));
        }
    }

    interface OnItemLongClickListener {
        boolean onItemLongClick(@NonNull HistoryItem item);
    }

    interface OnItemClickListener {
        void onItemClick(@NonNull HistoryItem item);
    }

    private static class BookmarkListAdapter extends RecyclerView.Adapter<BookmarkViewHolder> {

        @NonNull private List<HistoryItem> mBookmarks = new ArrayList<>();
        @NonNull private final FaviconModel mFaviconModel;
        @NonNull private final Bitmap mFolderBitmap;
        @NonNull private final Bitmap mWebpageBitmap;
        @NonNull private final Map<String, Subscription> mFaviconFetchSubscriptions = new ConcurrentHashMap<>();

        @Nullable private OnItemLongClickListener mOnItemLongCLickListener;
        @Nullable private OnItemClickListener mOnItemClickListener;

        BookmarkListAdapter(@NonNull FaviconModel faviconModel,
                            @NonNull Bitmap folderBitmap,
                            @NonNull Bitmap webpageBitmap) {
            mFaviconModel = faviconModel;
            mFolderBitmap = folderBitmap;
            mWebpageBitmap = webpageBitmap;
        }

        void setOnItemLongClickListener(@Nullable OnItemLongClickListener listener) {
            mOnItemLongCLickListener = listener;
        }

        void setOnItemClickListener(@Nullable OnItemClickListener listener) {
            mOnItemClickListener = listener;
        }

        @NonNull
        HistoryItem itemAt(int position) {
            return mBookmarks.get(position);
        }

        void deleteItem(@NonNull HistoryItem item) {
            List<HistoryItem> newList = new ArrayList<>(mBookmarks);
            newList.remove(item);
            updateItems(newList);
        }

        void updateItems(@NonNull List<HistoryItem> newList) {
            final List<HistoryItem> oldList = mBookmarks;
            mBookmarks = newList;

            DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new DiffUtil.Callback() {
                @Override
                public int getOldListSize() {
                    return oldList.size();
                }

                @Override
                public int getNewListSize() {
                    return mBookmarks.size();
                }

                @Override
                public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                    return oldList.get(oldItemPosition).equals(mBookmarks.get(newItemPosition));
                }

                @Override
                public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                    return oldList.get(oldItemPosition).equals(mBookmarks.get(newItemPosition));
                }
            });

            diffResult.dispatchUpdatesTo(this);
        }

        void cleanupSubscriptions() {
            for (Subscription subscription : mFaviconFetchSubscriptions.values()) {
                subscription.unsubscribe();
            }
            mFaviconFetchSubscriptions.clear();
        }

        @Override
        public BookmarkViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            LayoutInflater inflater = LayoutInflater.from(parent.getContext());
            View itemView = inflater.inflate(R.layout.bookmark_list_item, parent, false);

            return new BookmarkViewHolder(itemView, this, mOnItemLongCLickListener, mOnItemClickListener);
        }

        @Override
        public void onViewRecycled(BookmarkViewHolder holder) {
            super.onViewRecycled(holder);
        }

        @Override
        public void onBindViewHolder(final BookmarkViewHolder holder, int position) {
            ViewCompat.jumpDrawablesToCurrentState(holder.itemView);

            final HistoryItem web = mBookmarks.get(position);
            holder.txtTitle.setText(web.getTitle());
            if (web.isFolder()) {
                holder.favicon.setImageBitmap(mFolderBitmap);
            } else if (web.getBitmap() == null) {
                holder.favicon.setImageBitmap(mWebpageBitmap);
                holder.favicon.setTag(web.getUrl().hashCode());

                final String url = web.getUrl();

                Subscription oldSubscription = mFaviconFetchSubscriptions.get(url);
                SubscriptionUtils.safeUnsubscribe(oldSubscription);

                final Subscription faviconSubscription = mFaviconModel.faviconForUrl(url, web.getTitle())
                    .subscribeOn(Schedulers.io())
                    .observeOn(Schedulers.main())
                    .subscribe(new SingleOnSubscribe<Bitmap>() {
                        @Override
                        public void onItem(@Nullable Bitmap item) {
                            mFaviconFetchSubscriptions.remove(url);
                            Object tag = holder.favicon.getTag();
                            if (tag != null && tag.equals(url.hashCode())) {
                                holder.favicon.setImageBitmap(item);
                            }

                            web.setBitmap(item);
                        }
                    });

                mFaviconFetchSubscriptions.put(url, faviconSubscription);
            } else {
                holder.favicon.setImageBitmap(web.getBitmap());
            }

        }

        @Override
        public int getItemCount() {
            return mBookmarks.size();
        }
    }

}

File: app/src/main/java/acr/browser/lightning/browser/fragment/BookmarksFragment.kt
package acr.browser.lightning.browser.fragment

import acr.browser.lightning.R
import acr.browser.lightning.animation.AnimationUtils
import acr.browser.lightning.browser.BookmarksView
import acr.browser.lightning.browser.TabsManager
import acr.browser.lightning.browser.bookmark.BookmarkUiModel
import acr.browser.lightning.constant.LOAD_READING_URL
import acr.browser.lightning.controller.UIController
import acr.browser.lightning.database.Bookmark
import acr.browser.lightning.database.bookmark.BookmarkRepository
import acr.browser.lightning.di.DatabaseScheduler
import acr.browser.lightning.di.MainScheduler
import acr.browser.lightning.di.NetworkScheduler
import acr.browser.lightning.di.injector
import acr.browser.lightning.dialog.LightningDialogBuilder
import acr.browser.lightning.favicon.FaviconModel
import acr.browser.lightning.preference.UserPreferences
import acr.browser.lightning.reading.activity.ReadingActivity
import acr.browser.lightning.utils.ThemeUtils
import android.app.Activity
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.PorterDuff
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.IdRes
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import io.reactivex.Scheduler
import io.reactivex.Single
import io.reactivex.disposables.Disposable
import io.reactivex.rxkotlin.subscribeBy
import kotlinx.android.synthetic.main.bookmark_drawer.*
import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject

class BookmarksFragment : Fragment(), View.OnClickListener, View.OnLongClickListener, BookmarksView {

    // Managers
    @Inject internal lateinit var bookmarkModel: BookmarkRepository

    // Dialog builder
    @Inject internal lateinit var bookmarksDialogBuilder: LightningDialogBuilder

    @Inject internal lateinit var userPreferences: UserPreferences

    @Inject internal lateinit var faviconModel: FaviconModel

    @Inject @field:DatabaseScheduler internal lateinit var databaseScheduler: Scheduler
    @Inject @field:NetworkScheduler internal lateinit var networkScheduler: Scheduler
    @Inject @field:MainScheduler internal lateinit var mainScheduler: Scheduler

    private lateinit var uiController: UIController

    // Adapter
    private var bookmarkAdapter: BookmarkListAdapter? = null

    // Preloaded images
    private var webPageBitmap: Bitmap? = null
    private var folderBitmap: Bitmap? = null

    // Colors
    private var iconColor: Int = 0
    private var scrollIndex: Int = 0

    private var isIncognito: Boolean = false

    private var bookmarksSubscription: Disposable? = null
    private var bookmarkUpdateSubscription: Disposable? = null

    private val uiModel = BookmarkUiModel()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        injector.inject(this)

        val context = requireNotNull(context) { "Context should never be null in onCreate" }

        uiController = context as UIController
        isIncognito = arguments?.getBoolean(INCOGNITO_MODE, false) == true
        val darkTheme = userPreferences.useTheme != 0 || isIncognito
        webPageBitmap = ThemeUtils.getThemedBitmap(context, R.drawable.ic_webpage, darkTheme)
        folderBitmap = ThemeUtils.getThemedBitmap(context, R.drawable.ic_folder, darkTheme)
        iconColor = ThemeUtils.getIconThemeColor(context, darkTheme)
    }

    override fun onResume() {
        super.onResume()
        if (bookmarkAdapter != null) {
            setBookmarksShown(null, false)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View =
        inflater.inflate(R.layout.bookmark_drawer, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        bookmark_back_button_image?.setColorFilter(iconColor, PorterDuff.Mode.SRC_IN)
        val backView = view.findViewById<View>(R.id.bookmark_back_button)
        backView.setOnClickListener {
            if (!uiModel.isCurrentFolderRoot()) {
                setBookmarksShown(null, true)
                bookmark_list_view?.layoutManager?.scrollToPosition(scrollIndex)
            }
        }
        setupNavigationButton(view, R.id.action_add_bookmark, R.id.action_add_bookmark_image)
        setupNavigationButton(view, R.id.action_reading, R.id.action_reading_image)
        setupNavigationButton(view, R.id.action_toggle_desktop, R.id.action_toggle_desktop_image)


        bookmarkAdapter = BookmarkListAdapter(
            faviconModel,
            folderBitmap!!,
            webPageBitmap!!,
            networkScheduler,
            mainScheduler,
            this::handleItemLongPress,
            this::handleItemClick
        )

        bookmark_list_view?.let {
            it.layoutManager = LinearLayoutManager(context)
            it.adapter = bookmarkAdapter
        }

        setBookmarksShown(null, true)
    }

    override fun onDestroyView() {
        super.onDestroyView()

        bookmarksSubscription?.dispose()
        bookmarkUpdateSubscription?.dispose()

        bookmarkAdapter?.cleanupSubscriptions()
    }

    override fun onDestroy() {
        super.onDestroy()

        bookmarksSubscription?.dispose()
        bookmarkUpdateSubscription?.dispose()

        bookmarkAdapter?.cleanupSubscriptions()
    }

    private fun getTabsManager(): TabsManager = uiController.getTabModel()

    fun reinitializePreferences() {
        val activity = activity ?: return
        val darkTheme = userPreferences.useTheme != 0 || isIncognito
        webPageBitmap = ThemeUtils.getThemedBitmap(activity, R.drawable.ic_webpage, darkTheme)
        folderBitmap = ThemeUtils.getThemedBitmap(activity, R.drawable.ic_folder, darkTheme)
        iconColor = ThemeUtils.getIconThemeColor(activity, darkTheme)
    }

    private fun updateBookmarkIndicator(url: String) {
        bookmarkUpdateSubscription?.dispose()
        bookmarkUpdateSubscription = bookmarkModel.isBookmark(url)
            .subscribeOn(databaseScheduler)
            .observeOn(mainScheduler)
            .subscribe { boolean ->
                bookmarkUpdateSubscription = null
                val activity = activity
                if (action_add_bookmark_image == null || activity == null) {
                    return@subscribe
                }
                if (boolean) {
                    action_add_bookmark_image?.setImageResource(R.drawable.ic_bookmark)
                    action_add_bookmark_image?.setColorFilter(ThemeUtils.getAccentColor(activity), PorterDuff.Mode.SRC_IN)
                } else {
                    action_add_bookmark_image?.setImageResource(R.drawable.ic_action_star)
                    action_add_bookmark_image?.setColorFilter(iconColor, PorterDuff.Mode.SRC_IN)
                }
            }
    }

    override fun handleBookmarkDeleted(bookmark: Bookmark) = when (bookmark) {
        is Bookmark.Folder -> setBookmarksShown(null, false)
        is Bookmark.Entry -> bookmarkAdapter?.deleteItem(BookmarkViewModel(bookmark)) ?: Unit
    }

    private fun setBookmarksShown(folder: String?, animate: Boolean) {
        bookmarksSubscription?.dispose()
        bookmarksSubscription = bookmarkModel.getBookmarksFromFolderSorted(folder)
            .concatWith(Single.defer {
                if (folder == null) {
                    bookmarkModel.getFoldersSorted()
                } else {
                    Single.just(emptyList())
                }
            })
            .toList()
            .map { it.flatten() }
            .subscribeOn(databaseScheduler)
            .observeOn(mainScheduler)
            .subscribe { bookmarksAndFolders ->
                uiModel.currentFolder = folder
                setBookmarkDataSet(bookmarksAndFolders, animate)
            }
    }

    private fun setBookmarkDataSet(items: List<Bookmark>, animate: Boolean) {
        bookmarkAdapter?.updateItems(items.map { BookmarkViewModel(it) })
        val resource = if (uiModel.isCurrentFolderRoot()) {
            R.drawable.ic_action_star
        } else {
            R.drawable.ic_action_back
        }

        if (animate) {
            bookmark_back_button_image?.let {
                val transition = AnimationUtils.createRotationTransitionAnimation(it, resource)
                it.startAnimation(transition)
            }
        } else {
            bookmark_back_button_image?.setImageResource(resource)
        }
    }

    private fun setupNavigationButton(view: View, @IdRes buttonId: Int, @IdRes imageId: Int) {
        val frameButton = view.findViewById<FrameLayout>(buttonId)
        frameButton.setOnClickListener(this)
        frameButton.setOnLongClickListener(this)
        val buttonImage = view.findViewById<ImageView>(imageId)
        buttonImage.setColorFilter(iconColor, PorterDuff.Mode.SRC_IN)
    }

    private fun handleItemLongPress(bookmark: Bookmark): Boolean {
        (context as Activity?)?.let {
            when (bookmark) {
                is Bookmark.Folder -> bookmarksDialogBuilder.showBookmarkFolderLongPressedDialog(it, uiController, bookmark)
                is Bookmark.Entry -> bookmarksDialogBuilder.showLongPressedDialogForBookmarkUrl(it, uiController, bookmark)
            }
        }
        return true
    }

    private fun handleItemClick(bookmark: Bookmark) = when (bookmark) {
        is Bookmark.Folder -> {
            scrollIndex = (bookmark_list_view.layoutManager as LinearLayoutManager).findFirstVisibleItemPosition()
            setBookmarksShown(bookmark.title, true)
        }
        is Bookmark.Entry -> uiController.bookmarkItemClicked(bookmark)
    }


    override fun onClick(v: View) {
        when (v.id) {
            R.id.action_add_bookmark -> uiController.bookmarkButtonClicked()
            R.id.action_reading -> {
                val currentTab = getTabsManager().currentTab
                if (currentTab != null) {
                    val read = Intent(activity, ReadingActivity::class.java)
                    read.putExtra(LOAD_READING_URL, currentTab.url)
                    startActivity(read)
                }
            }
            R.id.action_toggle_desktop -> {
                getTabsManager().currentTab?.apply {
                    toggleDesktopUA()
                    reload()
                    // TODO add back drawer closing
                }
            }
            else -> {
            }
        }
    }

    override fun onLongClick(v: View) = false

    override fun navigateBack() {
        if (uiModel.isCurrentFolderRoot()) {
            uiController.onBackButtonPressed()
        } else {
            setBookmarksShown(null, true)
            bookmark_list_view?.layoutManager?.scrollToPosition(scrollIndex)
        }
    }

    override fun handleUpdatedUrl(url: String) {
        updateBookmarkIndicator(url)
        val folder = uiModel.currentFolder
        setBookmarksShown(folder, false)
    }

    private class BookmarkViewHolder(
        itemView: View,
        private val adapter: BookmarkListAdapter,
        private val onItemLongClickListener: (Bookmark) -> Boolean,
        private val onItemClickListener: (Bookmark) -> Unit
    ) : RecyclerView.ViewHolder(itemView), View.OnClickListener, View.OnLongClickListener {

        var txtTitle: TextView = itemView.findViewById(R.id.textBookmark)
        var favicon: ImageView = itemView.findViewById(R.id.faviconBookmark)

        init {
            itemView.setOnLongClickListener(this)
            itemView.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            val index = adapterPosition
            if (index.toLong() != RecyclerView.NO_ID) {
                onItemClickListener(adapter.itemAt(index).bookmark)
            }
        }

        override fun onLongClick(v: View): Boolean {
            val index = adapterPosition
            return index != RecyclerView.NO_POSITION && onItemLongClickListener(adapter.itemAt(index).bookmark)
        }
    }

    private class BookmarkListAdapter(
        private val faviconModel: FaviconModel,
        private val folderBitmap: Bitmap,
        private val webPageBitmap: Bitmap,
        private val networkScheduler: Scheduler,
        private val mainScheduler: Scheduler,
        private val onItemLongClickListener: (Bookmark) -> Boolean,
        private val onItemClickListener: (Bookmark) -> Unit
    ) : RecyclerView.Adapter<BookmarkViewHolder>() {

        private var bookmarks: List<BookmarkViewModel> = listOf()
        private val faviconFetchSubscriptions = ConcurrentHashMap<String, Disposable>()

        fun itemAt(position: Int): BookmarkViewModel = bookmarks[position]

        fun deleteItem(item: BookmarkViewModel) {
            val newList = bookmarks - item
            updateItems(newList)
        }

        fun updateItems(newList: List<BookmarkViewModel>) {
            val oldList = bookmarks
            bookmarks = newList

            val diffResult = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
                override fun getOldListSize() = oldList.size

                override fun getNewListSize() = bookmarks.size

                override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int) =
                    oldList[oldItemPosition] == bookmarks[newItemPosition]

                override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int) =
                    oldList[oldItemPosition] == bookmarks[newItemPosition]
            })

            diffResult.dispatchUpdatesTo(this)
        }

        fun cleanupSubscriptions() {
            for (subscription in faviconFetchSubscriptions.values) {
                subscription.dispose()
            }
            faviconFetchSubscriptions.clear()
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BookmarkViewHolder {
            val inflater = LayoutInflater.from(parent.context)
            val itemView = inflater.inflate(R.layout.bookmark_list_item, parent, false)

            return BookmarkViewHolder(itemView, this, onItemLongClickListener, onItemClickListener)
        }

        override fun onBindViewHolder(holder: BookmarkViewHolder, position: Int) {
            holder.itemView.jumpDrawablesToCurrentState()

            val viewModel = bookmarks[position]
            holder.txtTitle.text = viewModel.bookmark.title


            val bitmap = viewModel.icon ?: when (viewModel.bookmark) {
                is Bookmark.Folder -> folderBitmap
                is Bookmark.Entry -> webPageBitmap.also { _ ->
                    holder.favicon.tag = viewModel.bookmark.url.hashCode()

                    val url = viewModel.bookmark.url

                    faviconFetchSubscriptions[url]?.dispose()
                    faviconFetchSubscriptions[url] = faviconModel.faviconForUrl(url, viewModel.bookmark.title)
                        .subscribeOn(networkScheduler)
                        .observeOn(mainScheduler)
                        .subscribeBy(
                            onSuccess = {
                                viewModel.icon = it
                                if (holder.favicon.tag == url.hashCode()) {
                                    holder.favicon.setImageBitmap(it)
                                }
                            }
                        )
                }
            }

            holder.favicon.setImageBitmap(bitmap)

        }

        override fun getItemCount() = bookmarks.size
    }

    companion object {

        @JvmStatic
        fun createFragment(isIncognito: Boolean) = BookmarksFragment().apply {
            arguments = Bundle().apply {
                putBoolean(BookmarksFragment.INCOGNITO_MODE, isIncognito)
            }
        }

        private const val TAG = "BookmarksFragment"

        private const val INCOGNITO_MODE = "$TAG.INCOGNITO_MODE"
    }

}

private class BookmarkViewModel(
    val bookmark: Bookmark,
    var icon: Bitmap? = null
) {
    override fun equals(other: Any?): Boolean {
        return if (other is BookmarkViewModel) {
            bookmark == other.bookmark
        } else {
            super.equals(other)
        }
    }

    override fun hashCode(): Int = bookmark.hashCode()

    override fun toString(): String = "BookmarkViewModel(bookmark=$bookmark)"

}
